### PR TITLE
add support for mem_banks for --add-pskernel

### PR DESCRIPTION
**Pull request by xfreid**
#### Problem solved by the commit
Add support for mem_banks for --add-pskernel

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
N/A

#### How problem was solved, alternative solutions (if any) and why they were rejected
N/A

#### Risks (if any) associated the changes in the commit
None

#### What has been tested and how, request additional testing if necessary
Manual and unit test

#### Documentation impact (if any)
None

### CODE REVIEW

**Comment by xfreid** [comment]
Currently, the output looks like the following
`Specifying memory banks in --add-pskernel is an adanced feature
	connecting arg 0 to mem bank 0
	connecting arg 0 to mem bank 1
	connecting arg 1 to mem bank 0
	connecting arg 1 to mem bank 1
	connecting arg 2 to mem bank 0
	connecting arg 2 to mem bank 1
	connecting arg 6 to mem bank 0
	connecting arg 6 to mem bank 1
`

**Comment by xfreid** [comment]
retest this please

**Comment by dayeh-xilinx** [comment]
retest this please

File: src/runtime_src/tools/xclbinutil/KernelUtilities.h [hunk 1]
```diff
@@ -26,6 +26,6 @@ namespace XclBinUtilities {
 void addKernel(const boost::property_tree::ptree& ptKernel, bool isFixedPS, boost::property_tree::ptree& ptEmbeddedData);
```
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/KernelUtilities.h:26]: Update copyright date on line 2.

File: src/runtime_src/tools/xclbinutil/XclBinUtilMain.cxx [hunk 2]
```diff
@@ -165,7 +165,7 @@ int main_(int argc, const char** argv) {
   po::options_description desc("Options");
   desc.add_options()
       ("add-merge-section", boost::program_options::value<decltype(sectionsToAddMerge)>(&sectionsToAddMerge)->multitoken(), "Section name to add or merge.  Format: <section>:<format>:<file>")
-      ("add-pskernel", boost::program_options::value<decltype(addPsKernels)>(&addPsKernels)->multitoken(), "Helper option to add PS kernels.  Format: <symbol_name>:<instances>:<path_to_shared_library>")
+      ("add-pskernel", boost::program_options::value<decltype(addPsKernels)>(&addPsKernels)->multitoken(), "Helper option to add PS kernels.  Format: [<mem_banks>]:[<symbol_name>]:[<instances>]:<path_to_shared_library>")
```
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/XclBinUtilMain.cxx:168]: Nice.  Thank you it is cleaner.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 3]
```diff
@@ -1723,7 +1724,18 @@ parsePSKernelString(const std::string& encodedString,
...
-  XUtil::TRACE(boost::format("PSKernel command arguments: symbol_name='%s'; num_instances=%d; library='%s'") % symbol_name % num_instances % path_to_library);
+  // -- [3]: Mem banks --
+  mem_banks = (tokens.size() > 3) ? tokens[3] : "";
+
+  // add check for leading and trailing ;
+  if (!mem_banks.empty()) {
+    if (mem_banks.front() == ',' || mem_banks.back() == ',' )
+      throw std::runtime_error("Specifїed mem_banks is not valid");
+    
+    std::cout << "Specifying memory banks in --add-pskernel is an adanced feature" << std::endl;
```
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1735]:
```suggestion
    std::cout << "Attention: Specifying memory banks in --add-pskernel is an advanced feature." << std::endl;
    std::cout << "                 Be sure to validate connections after performing this operation." << std::endl;
```

File: src/runtime_src/tools/xclbinutil/KernelUtilities.cxx [hunk 4]
```diff
@@ -306,26 +322,67 @@ void addArgsToMemoryConnections(const unsigned int ipLayoutIndexID,
...
+
+      if (isNumeric(memoryConnection)) {
+        // make sure the specified index exists
+        for (const auto& memBankIndex : viMemBanks) {
+          if (memBankIndex >= 0 && memBankIndex < static_cast<int>(memTopology.size())) {
+            vMemBanks.push_back(std::to_string(memBankIndex));
+            std::cout << "\tconnecting arg " << std::to_string(argIndexID) << " to mem bank " << std::to_string(memBankIndex) << std::endl;
+          }
+          else
+            throw std::runtime_error("Specifїed mem_banks is not valid");
```
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/KernelUtilities.cxx:358]: This error message should be a bit more descriptive.  For example:
```suggestion
         {
            const auto errMsg = boost::format("Specified memory bank (%d) is invalid.  Valid ranges are from 0 to %d.") % memBankIndex % memTopology.size(); 
            throw std::runtime_error(errMsg.str());
         }
```

**Review by rozumx — CHANGES_REQUESTED** [review]
Some minor changes are being requested.

File: src/runtime_src/tools/xclbinutil/KernelUtilities.h (hunk 1 above)
Comment (xfreid) [reply src/runtime_src/tools/xclbinutil/KernelUtilities.h:27]: copyright added

File: src/runtime_src/tools/xclbinutil/unittests/PSKernel/PSKernel.py [hunk 5]
```diff
@@ -96,7 +95,7 @@ def main():
                      "--dump-section", "CONNECTIVITY:JSON:" + outputConnectivity,
                      "--dump-section", "MEM_TOPOLOGY:JSON:" + outputMemTopology,
                      "--output", outputXCLBIN, 
-                     "--force"
+                     "--force", "--trace"
```
[the remark this answers is not in the document]
Comment (xfreid) [reply src/runtime_src/tools/xclbinutil/unittests/PSKernel/PSKernel.py]: --trace removed

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 3 above)
Comment (xfreid) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx]: message updated

File: src/runtime_src/tools/xclbinutil/KernelUtilities.cxx (hunk 4 above)
Comment (xfreid) [reply src/runtime_src/tools/xclbinutil/KernelUtilities.cxx]: message udpated

**Comment by xfreid** [comment]
retest this please

**Comment by manikandan-xilinx** [comment]
retest this please.

**Comment by xfreid** [comment]
retest this please

**Review by rozumx — APPROVED** [review]
Thank you for making these changes.

**Comment by rozumx** [comment]
retest this please

**Comment by manikandan-xilinx** [comment]
retest this please.

**Pull request merged**